File: backend/src/auth/auth.controller.ts
import { Body, Controller, HttpCode, Param, Post, Req, Res, UseGuards, UsePipes, ValidationPipe } from '@nestjs/common';
import { Request, Response } from 'express';
import { AuthService } from './auth.service';
import { AuthDto } from './dto/auth.dto';
import { JwtAuthGuard } from './token/access-token.guard';

@Controller('auth')
export class AuthController {
  private AGE_30D = 30 * 24 * 60 * 60 * 1000

  constructor(private readonly authService: AuthService) {}

  @UsePipes(new ValidationPipe())
  @HttpCode(200)
  @Post('login')
  async login(@Body() dto: AuthDto, @Res({ passthrough: true }) response: Response) {
    const result = await this.authService.login(dto)
    this.saveRefreshTokenToCookie(response, result.refreshToken)
    return result
  }

  @UsePipes(new ValidationPipe())
  @Post('registration')
  @HttpCode(200)
  async registration(@Body() dto: AuthDto, @Res({ passthrough: true }) response: Response) {
    const result = await this.authService.registration(dto)
    this.saveRefreshTokenToCookie(response, result.refreshToken)
    return result
  }
  
  @Post('logout')
  @HttpCode(200)
  async logout(@Req() request: Request, @Res({ passthrough: true }) response: Response) {
    const { refreshToken } = request.cookies
    await this.authService.logout(refreshToken)
    response.clearCookie('refreshToken')
  }

  @UseGuards(JwtAuthGuard)
  @Post('activate/:code')
  @HttpCode(200)
  async activate(@Param('code') code: number, @Req() request: Request) {
    await this.authService.activate(request.user['_id'], code)
  }

  @Post('refresh')
  async refresh(@Req() request: Request, @Res({ passthrough: true }) response: Response) {
    const { refreshToken } = request.cookies
    const result = await this.authService.refresh(refreshToken)
    this.saveRefreshTokenToCookie(response, result.refreshToken)
    return result
  }
  
  private saveRefreshTokenToCookie(response: Response, refreshToken: string) {
    response.cookie('refreshToken', refreshToken, { maxAge: this.AGE_30D, httpOnly: true });
  }
}

File: backend/Backend.Api/Controllers/User/AuthController.cs
using backend.Constants;
using Backend.Core.DTOs.User.Auth;
using Backend.Core.Services;
using Backend.Core.Services.User;
using backend.Extensions;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace backend.Controllers.User;

[Route("api/auth")]
public class AuthController : Controller
{
    private readonly AuthService _authService;

    public AuthController(AuthService authService)
    {
        _authService = authService;
    }

    [HttpPost("login")]
    public async Task<SuccessAuthDto> Login([FromBody] LoginDto dto)
    {
        var result = await _authService.Login(dto);
        AddRefreshTokenToCookie(result);
        return result;
    }
    
    [HttpPost("registration-dropshipper")]
    public async Task<SuccessAuthDto> RegistrationDropshipper([FromBody] RegistrationDropshipperDto dto)
    {
        var result = await _authService.RegistrationDropshipper(dto);
        AddRefreshTokenToCookie(result);
        return result;
    }
    
    [HttpPost("registration-supplier")]
    public async Task<SuccessAuthDto> RegistrationSupplier([FromBody] RegistrationSupplierDto dto)
    {
        var result = await _authService.RegistrationSupplier(dto);
        AddRefreshTokenToCookie(result);
        return result;
    }
    
    [HttpPost("refresh")]
    public async Task<SuccessAuthDto> Refresh()
    {
        if (Request.Cookies[CookieKeys.RefreshToken] == null)
            throw new BadHttpRequestException("Refresh token not found");

        var result = await _authService.Refresh(Request.Cookies[CookieKeys.RefreshToken]!);
        AddRefreshTokenToCookie(result);
        return result;
    }

    [Authorize]
    [HttpPost("activate/{confirmationCode}")]
    public async Task ConfirmEmail(int confirmationCode)
    {
        await _authService.ConfirmEmail(User.Identity!.GetId(), confirmationCode);
    }

    [Authorize]
    [HttpPost("logout")]
    public async Task Logout()
    {
        Response.Cookies.Delete(CookieKeys.RefreshToken);
        await Task.CompletedTask;
    }
    
    private void AddRefreshTokenToCookie(SuccessAuthDto result)
    {
        Response.Cookies.Append(CookieKeys.RefreshToken, result.RefreshToken, new CookieOptions
        {
            Expires = DateTime.Now.AddDays(30),
            HttpOnly = true,
            Path = "/"
        });
    }
}